### PR TITLE
image: hash inputs to image

### PR DESCRIPTION
**Pull request by jazzdan**
only return image from history if service
 hash matches

~I'm not happy with how many arguments we're passing to things, and I'm not convinced this is the right API. I welcome all feedback on how to make this a bit more clean.~ -I feel better now, thanks Nick!

### CODE REVIEW

**Comment by jazzdan** [comment]
@nicks that's a good point, and yeah I looked at it. We don't want to invalidate the cache on everything that a `Service` has, though I suppose we could pass that around and just grab the things we need. 

Unless anyone has other ideas I'll do that when I get back to this.

File: internal/image/history.go [hunk 1]
```diff
@@ -89,25 +93,91 @@ func (h ImageHistory) load(ctx context.Context, name reference.Named, digest dig
...
+	Mounts         []model.Mount
+	Steps          []model.Cmd
+	Entrypoint     model.Cmd
 }
 
-func (h ImageHistory) AddAndPersist(ctx context.Context, name reference.Named, digest digest.Digest, checkpoint CheckpointID) error {
-	key, entry := h.load(ctx, name, digest, checkpoint)
+type HashedInputs = uint64
+
+func hashInputs(service model.Service) (HashedInputs, error) {
```
Comment (jazzdan) [inline internal/image/history.go]: Will change this function name to `hashService`

File: internal/image/history.go [hunk 2]
```diff
@@ -89,25 +93,91 @@ func (h ImageHistory) load(ctx context.Context, name reference.Named, digest dig
...
+	Entrypoint     model.Cmd
 }
 
-func (h ImageHistory) AddAndPersist(ctx context.Context, name reference.Named, digest digest.Digest, checkpoint CheckpointID) error {
-	key, entry := h.load(ctx, name, digest, checkpoint)
+type HashedService = uint64
+
+func hashService(service model.Service) (HashedService, error) {
+	hi := hash{
+		BaseDockerfile: build.Dockerfile(service.DockerfileText),
```
Comment (nicks) [inline internal/image/history.go]: this is a dangerous approach. if someone adds new fields to Service, they won't know to update them here.

can you either:
1) hash the existing Service structure, but void out the fields you don't care about?
2) create some sort of hash method on Service itself?

File: internal/image/history.go [hunk 3]
```diff
@@ -89,25 +93,91 @@ func (h ImageHistory) load(ctx context.Context, name reference.Named, digest dig
 		h.byName[key] = bucket
 	}
 
+	bucket.entries = append(bucket.entries, entry)
+	if entry.After(bucket.mostRecent.CheckpointID) {
+		bucket.mostRecent = entry
+	}
+}
+
+func (h ImageHistory) load(
```
Comment (maiamcc) [inline internal/image/history.go]: is this replicating logic from `loadFromEntry`? maybe this should just assemble all the inputs into a `historyEntry` and call `loadFromEntry` on it?

File: internal/image/history.go (hunk 2 above)
Comment (maiamcc) [reply internal/image/history.go]: ooohhh hash method on the service++

Comment (jazzdan) [reply internal/image/history.go]: yeah that's a great point @nicks, I like that idea.

File: internal/model/service.go [hunk 4]
```diff
@@ -4,20 +4,40 @@ import (
...
+	"github.com/mitchellh/hashstructure"
 )
 
 type ServiceName string
 
 func (s ServiceName) String() string { return string(s) }
 
 type Service struct {
-	K8sYaml        string
+	K8sYaml        string `hash:"ignore"`
```
Comment (jazzdan) [inline internal/model/service.go]: Check this out @nicks. I didn't realize this was a feature of the library!

File: internal/image/history.go [hunk 5]
```diff
@@ -89,25 +117,45 @@ func (h ImageHistory) load(ctx context.Context, name reference.Named, digest dig
 		h.byName[key] = bucket
 	}
 
+	hi, err := service.Hash()
```
Comment (maiamcc) [inline internal/image/history.go]: 👋

File: internal/image/history.go [hunk 6]
```diff
@@ -117,7 +165,16 @@ func (h ImageHistory) MostRecent(name reference.Named) (digest.Digest, Checkpoin
 		return "", CheckpointID{}, false
 	}
 
+	hi, err := service.Hash()
```
Comment (maiamcc) [inline internal/image/history.go]: 👋

File: internal/model/service.go [hunk 7]
```diff
@@ -4,20 +4,40 @@ import (
...
 	DockerfileText string
 	Mounts         []Mount
 	Steps          []Cmd
 	Entrypoint     Cmd
```
Comment (maiamcc) [inline internal/model/service.go]: hmm I'm not conviiiinced we need to enforce ALL of these being constant to use an image for the same service... off the top of my head i knew we def want dockerfile, we might want entrypoint for incremental build, and idk if we need mounts or steps...

File: internal/model/service.go [hunk 8]
```diff
@@ -4,20 +4,40 @@ import (
...
-	DockerfileTag  string
-	Name           ServiceName
+	DockerfileTag  string      `hash:"ignore"`
+	Name           ServiceName `hash:"ignore"`
+}
+
+type HashedService = uint64
+
+type hash struct {
```
Comment (maiamcc) [inline internal/model/service.go]: is this still being used?

File: internal/image/history.go [hunk 9]
```diff
@@ -101,8 +99,8 @@ func (h ImageHistory) loadFromEntry(ctx context.Context, name reference.Named, e
 	}
 }
 
-// load takes checkpoint and a service definition and loads it in to the appropriate bucket in memory.
-func (h ImageHistory) load(
+// addInMemory takes checkpoint and a service definition and loads it in to the appropriate bucket in memory.
+func (h ImageHistory) addInMemory(
```
[the remark this answers is not in the document]
Comment (jazzdan) [reply internal/image/history.go]: oh yes thanks for the reminder

File: internal/model/service.go (hunk 7 above)
Comment (jazzdan) [reply internal/model/service.go]: I'll adjust this in the next step where we actually make use of this code.

File: internal/model/service.go (hunk 4 above)
Comment (nicks) [reply internal/model/service.go]: :raised_hands:

**Pull request merged**